### PR TITLE
[MODULAR] Allows emotes to be done using the container of a mob, AKA mech emote PR

### PR DESCRIPTION
**Pull request by nikothedude**
## About The Pull Request
Allows mobs within certain containers (e.g. mechs) to emote USING those containers with a seperate emote.

## How This Contributes To The Skyrat Roleplay Experience
You can maintain TOTAL anonymity in mechs now. Also I can punt small people with my mech now, so thats fun.
image

## Changelog

:cl:
add: Added a new verb that allows you to emote using the container your mob is in, using their name. The verb's name is *exme and either be used using the say window, command bar, or the IC tab as "Emote using Vehicle/Container".
/:cl:

## Progress
- [x] Proof of concept
- [x] Tidy up code (it wasnt that bad actually?)
- [x] Runetext integration
- [x] TEST IT (a little)

### CODE REVIEW

**Comment by nikothedude** [comment]
ANd yes this will be done upstream later, but theres a feature freeze right now and I'm practicing emote code

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	key_third_person = "container_emote"
+	message = null
+
+/datum/emote/container_emote/run_emote(mob/user, params, type_override = null)
+	var/container_message
+	var/container_emote = params
+	if(is_banned_from(user, "emote"))
+		to_chat(user, "You cannot send emotes (banned).")
+		return FALSE
+	else if(user.client && user.client.prefs.muted & MUTE_IC)
```
Comment (tf-4) [inline modular_skyrat/modules/container_emotes/code/container_emotes.dm]:
```suggestion
	else if(user.client?.prefs?.muted & MUTE_IC)
```

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		to_chat(user, "You cannot send IC messages (muted).")
+		return FALSE
+	else if(QDELETED(user))
+		return FALSE
+	else if (isturf(user.loc))
+		return FALSE
+	else if(!params)
+		container_emote = stripped_multiline_input(user, "Choose an emote to display.", "Container" , null, MAX_MESSAGE_LEN)
+		if(container_emote)
```
Comment (tf-4) [inline modular_skyrat/modules/container_emotes/code/container_emotes.dm]: Early return

**Review by tf-4 — COMMENTED** [review]
Great idea. Definitely take upstream when their freeze is done

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm (hunk 2 above)
Comment (nikothedude) [reply modular_skyrat/modules/container_emotes/code/container_emotes.dm]: Yeaaah, I copied most of this code from subtler code, soooo I wasn't expecting the best quality of the bat

Comment (nikothedude) [reply modular_skyrat/modules/container_emotes/code/container_emotes.dm]: Also, how is this an early return? It only returns false if the mob's location is a turf.

**Comment by Stalkeros2** [comment]
Mecha ERP. :flushed:

**Comment by nikothedude** [comment]
> Mecha ERP. 😳

SADLY this PR allows MANY FORMS OF ERP NOW. MANY.

**Comment by nikothedude** [comment]
image
kinda proud of myself tbh

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm (hunk 2 above)
Comment (nikothedude) [reply modular_skyrat/modules/container_emotes/code/container_emotes.dm]: nvm

**Comment by nikothedude** [comment]
I'm doing distance stuff in another PR tomorrow I think, mech subtle will come later

**Comment by nikothedude** [comment]
I've realized there isnt really much testing to do, soooooooo

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		to_chat(user, "You cannot send IC messages (muted).")
+		return FALSE
+	else if (isturf(user.loc))
+		to_chat(user, "You are not within anything!")
+		return FALSE
+	else if(!params)
+		container_emote = stripped_multiline_input(user, "Choose an emote to display.", "Container Emote" , null, MAX_MESSAGE_LEN)
+		if(!container_emote)
+			return FALSE
+		var/type = input("Is this a visible or hearable emote?") as null|anything in list("Visible", "Hearable")
```
Comment (jjpark-kb) [inline modular_skyrat/modules/container_emotes/code/container_emotes.dm]: instead of input, use tgui_input_list or something along those lines.

Comment (tf-4) [inline modular_skyrat/modules/container_emotes/code/container_emotes.dm]: why hearable and not just audible

Comment (nikothedude) [reply modular_skyrat/modules/container_emotes/code/container_emotes.dm]: Because uhhhhhhhhh

(this is copy and pasted)

**Comment by gonenoculer5** [comment]
@nikothedude Maybe make this work similar to how goldens PR fixing the whitespace between the person commiting the action and the actual content of the emote so its consistant.

**Comment by gonenoculer5** [comment]
https://gyazo.com/5fd0ae674073ca536b3c56e0958a2e97

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	else if (emote_type == EMOTE_VISIBLE)
+		user.loc.visible_message(message = container_message, self_message = container_message, visible_message_flags = EMOTE_MESSAGE, separation = space)
+
+
+
+
+
+
+
+
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/container_emotes/code/container_emotes.dm]: what is this? why is so many spaces? this is haram

**Comment by nikothedude** [comment]
okay okay im going to work on this again

**Comment by nikothedude** [comment]
> @nikothedude Maybe make this work similar to how goldens PR fixing the whitespace between the person commiting the action and the actual content of the emote so its consistant.

did that already

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm (hunk 4 above)
Comment (nikothedude) [reply modular_skyrat/modules/container_emotes/code/container_emotes.dm]: this is not to be questioned

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	if(!can_run_emote(user))
+		return FALSE
+
+	user.log_message(container_message, LOG_EMOTE)
+
+	var/space = should_have_space_before_emote(html_decode(container_emote)[1]) ? " " : ""
+
+	container_message = ("[user.say_emphasis(container_message)]")
+
+	if (isturf(user.loc) || (!user.loc) || user.IsUnconscious()) //one last sanity check
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/container_emotes/code/container_emotes.dm:63]: Good call there for that sanity check :+1:

File: modular_skyrat/modules/container_emotes/readme.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+<!-- Any master file changes you've made to existing master files or if you've added a new master file. Please mark either as #NEW or #CHANGE -->
+
+### Included files that are not contained in this module:
+
+- N/A
+<!-- Likewise, be it a non-modular file or a modular one that's not contained within the folder belonging to this specific module, it should be mentioned here -->
+
+### Credits:
+- Niko#7526
+<!-- Here go the credits to you, dear coder, and in case of collaborative work or ports, credits to the original source of the code -->
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/container_emotes/readme.md]: No need to keep all the comments :)
```suggestion
### TG Proc/File Changes:
- N/A

### Defines:
- N/A

### Master file additions
- N/A

### Included files that are not contained in this module:
- N/A

### Credits:
- Niko#7526
```

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+			return FALSE
+		var/list/choices = list("Visible","Audible")
+		var/type = tgui_input_list(user, "Is this a visible or audible emote?", "Container Emote", choices, 0)
+		switch(type)
+			if("Visible")
+				emote_type = EMOTE_VISIBLE
+			if("Audible")
+				emote_type = EMOTE_AUDIBLE
+			else
+				alert("Unable to use this emote, must be either audible or visible.")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/container_emotes/code/container_emotes.dm]: I think there's TGUI alerts, isn't there?

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
Good work so far, definitely would test-merge this first though.

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm (hunk 7 above)
Comment (nikothedude) [reply modular_skyrat/modules/container_emotes/code/container_emotes.dm]: Maybe, I'll check

Comment (nikothedude) [reply modular_skyrat/modules/container_emotes/code/container_emotes.dm]: You know what, fuck it, I'll make it a bubble alert, we love bubble alerts here :)

File: modular_skyrat/modules/container_emotes/code/container_emotes.dm [hunk 8]
```diff
@@ -39,11 +41,11 @@
 			if("Audible")
 				emote_type = EMOTE_AUDIBLE
 			else
-				alert("Unable to use this emote, must be either audible or visible.")
+				user.balloon_alert(user, "Unable to use this emote, must be either audible or visible.")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/container_emotes/code/container_emotes.dm]: I wouldn't use a balloon alert, I meant to use the `tgui_alert()` or whatever it's actually called, I forgot. The thing that you just click "ok"

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
This is nice and all, but you should mention it by name in the changelog too. Otherwise, it's ready to merge.

**Review by GoldenAlpharex — APPROVED** [review]

**Pull request merged**